File: user-service/src/main/java/com/github/manosbatsis/services/user/kafka/UserStream.java
package com.github.manosbatsis.services.user.kafka;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.manosbatsis.services.user.exception.UserStreamJsonProcessingException;
import com.github.manosbatsis.services.user.messages.EventType;
import com.github.manosbatsis.services.user.messages.UserEventMessage;
import com.github.manosbatsis.services.user.model.User;
import com.github.manosbatsis.services.user.rest.dto.CreateUserRequest;
import com.github.manosbatsis.services.user.rest.dto.UpdateUserRequest;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cloud.stream.function.StreamBridge;
import org.springframework.messaging.Message;
import org.springframework.messaging.support.MessageBuilder;
import org.springframework.stereotype.Component;
import org.springframework.util.MimeType;

import java.util.UUID;

@Slf4j
@Component
public class UserStream {

    @Autowired
    private StreamBridge streamBridge;
    @Autowired
    private ObjectMapper objectMapper;

    @Value("${spring.cloud.stream.bindings.users-out-0.content-type:application/json}")
    private String streamOutMimeType;

    public Message<UserEventMessage> userCreated(Long userId, CreateUserRequest createUserRequest) {
        return sendToBus(userId, EventType.CREATED, writeValueAsString(createUserRequest));
    }

    public Message<UserEventMessage> userCreated(User user) {
        return sendToBus(user.getId(), EventType.CREATED, writeValueAsString(user));
    }

    public Message<UserEventMessage> userUpdated(Long userId, UpdateUserRequest updateUserRequest) {
        return sendToBus(userId, EventType.UPDATED, writeValueAsString(updateUserRequest));
    }

    public Message<UserEventMessage> userUpdated(User user) {
        return sendToBus(user.getId(), EventType.UPDATED, writeValueAsString(user));
    }

    public Message<UserEventMessage> userDeleted(Long userId) {
        return sendToBus(userId, EventType.DELETED, null);
    }

    private Message<UserEventMessage> sendToBus(Long userId, EventType eventType, String json) {
        UserEventMessage userEventMessage = new UserEventMessage(
            UUID.randomUUID().toString(),
            System.currentTimeMillis(),
            eventType,
            userId,
            json
        );

        return sendToBus(userId, userEventMessage);
    }

    private Message<UserEventMessage> sendToBus(Long partitionKey, UserEventMessage userEventMessage) {
        Message<UserEventMessage> message = MessageBuilder.withPayload(userEventMessage)
            .setHeader("partitionKey", partitionKey)
            .setHeader("target-protocol", "kafka")
            .build();

        streamBridge.send("users-topic", message, MimeType.valueOf(streamOutMimeType));
        return message;
    }


    private String writeValueAsString(Object object) {
        try {
            return objectMapper.writeValueAsString(object);
        } catch (JsonProcessingException e) {
            throw new UserStreamJsonProcessingException(e);
        }
    }
}

File: user-service/src/main/kotlin/com/github/manosbatsis/services/user/kafka/UserStream.kt
package com.github.manosbatsis.services.user.kafka

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.github.manosbatsis.lib.core.log.loggerFor
import com.github.manosbatsis.services.user.exception.UserStreamJsonProcessingException
import com.github.manosbatsis.services.user.messages.EventType
import com.github.manosbatsis.services.user.messages.UserEventMessage
import com.github.manosbatsis.services.user.model.User
import com.github.manosbatsis.services.user.rest.dto.CreateUserRequest
import com.github.manosbatsis.services.user.rest.dto.UpdateUserRequest
import org.springframework.beans.factory.annotation.Value
import org.springframework.cloud.stream.function.StreamBridge
import org.springframework.messaging.Message
import org.springframework.messaging.support.MessageBuilder
import org.springframework.stereotype.Component
import org.springframework.util.MimeType
import java.util.UUID

@Component
class UserStream(
    val streamBridge: StreamBridge,
    val objectMapper: ObjectMapper,
) {
    companion object {
        private val log = loggerFor<UserStream>()
    }

    @Value("\${spring.cloud.stream.bindings.users-out-0.content-type:application/json}")
    private lateinit var streamOutMimeType: String

    fun userCreated(userId: Long, createUserRequest: CreateUserRequest): Message<UserEventMessage> {
        return sendToBus(userId, EventType.CREATED, writeValueAsString(createUserRequest))
    }

    fun userCreated(user: User): Message<UserEventMessage> {
        return sendToBus(user.id, EventType.CREATED, writeValueAsString(user))
    }

    fun userUpdated(userId: Long, updateUserRequest: UpdateUserRequest): Message<UserEventMessage> {
        return sendToBus(userId, EventType.UPDATED, writeValueAsString(updateUserRequest))
    }

    fun userUpdated(user: User): Message<UserEventMessage> {
        return sendToBus(user.id, EventType.UPDATED, writeValueAsString(user))
    }

    fun userDeleted(user: User): Message<UserEventMessage> {
        return sendToBus(user.id, EventType.DELETED, writeValueAsString(user))
    }

    private fun sendToBus(userId: Long, eventType: EventType, json: String?): Message<UserEventMessage> {
        val userEventMessage = UserEventMessage(
            UUID.randomUUID().toString(),
            System.currentTimeMillis(),
            eventType,
            userId,
            json,
        )
        return sendToBus(userId, userEventMessage)
    }

    private fun sendToBus(
        partitionKey: Long,
        userEventMessage: UserEventMessage,
    ): Message<UserEventMessage> {
        log.info("sendToBus, userId: {}, eventType: {}", userEventMessage.userId, userEventMessage.eventType)
        val message = MessageBuilder.withPayload(userEventMessage)
            .setHeader("partitionKey", partitionKey)
            .setHeader("target-protocol", "kafka")
            .build()
        streamBridge.send("users-topic", message, MimeType.valueOf(streamOutMimeType))
        return message
    }

    private fun writeValueAsString(`object`: Any): String {
        return try {
            objectMapper.writeValueAsString(`object`)
        } catch (e: JsonProcessingException) {
            throw UserStreamJsonProcessingException(e)
        }
    }
}
